src/rust/engine/ui/src/console_ui.rs

```diff
@@ -0,0 +1,222 @@
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::future::{self, FutureExt, TryFutureExt};
+use indexmap::IndexMap;
+use indicatif::{MultiProgress, ProgressBar, ProgressDrawTarget, ProgressStyle};
+use uuid::Uuid;
+
+use logging::logger::{StdioHandler, LOGGER};
+use task_executor::Executor;
+use workunit_store::WorkUnitStore;
+
+pub struct ConsoleUI {
+  workunit_store: WorkUnitStore,
+  // While the UI is running, there will be an Instance present.
+  instance: Option<Instance>,
+}
+
+impl ConsoleUI {
+  pub fn new(workunit_store: WorkUnitStore) -> ConsoleUI {
+    ConsoleUI {
+      workunit_store,
+      instance: None,
+    }
+  }
+
+  fn default_draw_target() -> ProgressDrawTarget {
+    // NB: We render more frequently than we receive new data in order to minimize aliasing where a
+    // render might barely miss a data refresh.
+    ProgressDrawTarget::stderr_with_hz(Self::render_rate_hz() * 2)
+  }
+
+  ///
+  /// The number of times per-second that `Self::render` should be called.
+  ///
+  pub fn render_rate_hz() -> u64 {
+    10
+  }
+
+  pub fn write_stdout(&self, msg: &str) {
+    print!("{}", msg);
+  }
+
+  pub fn write_stderr(&self, msg: &str) {
+    if let Some(instance) = &self.instance {
+      instance.bars[0].println(msg);
+    } else {
+      eprintln!("{}", msg);
+    }
+  }
+
+  pub fn with_console_ui_disabled<F: FnOnce() -> T, T>(&self, f: F) -> T {
+    if let Some(instance) = &self.instance {
+      instance
+        .multi_progress
+        .set_draw_target(ProgressDrawTarget::hidden());
+      let res = f();
+      instance
+        .multi_progress
+        .set_draw_target(Self::default_draw_target());
+      res
+    } else {
+      f()
+    }
+  }
+
+  fn setup_bars(num_swimlanes: usize) -> (MultiProgress, Vec<ProgressBar>) {
+    let multi_progress_bars = MultiProgress::with_draw_target(Self::default_draw_target());
+
+    let bars = (0..num_swimlanes)
+      .map(|_n| {
+        let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+        multi_progress_bars.add(ProgressBar::new(50).with_style(style))
+      })
+      .collect();
+
+    (multi_progress_bars, bars)
+  }
+
+  fn get_label_from_heavy_hitters<'a>(
+    tasks_to_display: impl Iterator<Item = (&'a String, &'a Option<Duration>)>,
+  ) -> Vec<String> {
+    tasks_to_display
+      .map(|(label, maybe_duration)| {
+        let duration_label = match maybe_duration {
+          None => "(Waiting) ".to_string(),
```

Comment:
Thoughts on `(Queued)`? It's a more active and positive verb imo. In my head, rather than "I'm waiting...waiting..waiting" like an impatient kid, it's closer to "Oh, cool, Pants queues stuff."

Reply:
Hm. As previously discussed, "queued" or "waiting" things ever showing up in the UI would be a bug. If there are spare swimlanes, then there should also be spare slots on the CommandRunner (because the number of slots should be aligned).